### PR TITLE
Convert generator in Sampler back to lazy construction

### LINKED ISSUE #63609: Sampler should be seeded lazily
## 🐛 Bug

After #63026 is landed, the generator for Sampler is attached to the instance, which helps to serialize the state of Sampler. But, it brings a problem that it will prevent Sampler's generator being seeded before each epoch.

## To Reproduce

Check https://github.com/pytorch/pytorch/pull/63026#issuecomment-902234490

### PR DESCRIPTION
**Pull request by ejguan**
Fixes #63609

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65857
* __->__ #63646

- Revert #63026 
  - Sampler is expected to be re-seeded if user specify seed before each epoch
  - Can not attach generator to self with `__iter__` because multiple iterators will ruin the use case
- Add tests to prevent the same case for different Samplers

Differential Revision: [D30451774](https://our.internmc.facebook.com/intern/diff/D30451774)

### CODE REVIEW

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

File: torch/utils/data/sampler.py [hunk 1]
```diff
@@ -89,6 +89,7 @@ def __init__(self, data_source: Sized, replacement: bool = False,
         self.replacement = replacement
         self._num_samples = num_samples
         self.generator = generator
+        self.gen: Optional[torch.Generator] = None
```
Comment (vfdev-5) [inline torch/utils/data/sampler.py:92]: Maybe, `self.gen` can be a private attribute (no worries about later changing and BC etc) ?

**Review by vfdev-5 — COMMENTED** [review]
@ejguan looks good to me. Thanks. I left a nit comment.

File: torch/utils/data/sampler.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/sampler.py:92]: Makes sense to me.
BTW, adding `gen` to self may provide a way to resume in the middle of epoch by serializing/deserializing the state of `gen`, since the `gen` is attached to the sampler instance now.

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

File: torch/utils/data/sampler.py [hunk 2]
```diff
@@ -112,15 +113,19 @@ def num_samples(self) -> int:
...
         if self.replacement:
             for _ in range(self.num_samples // 32):
-                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=self.generator).tolist()
-            yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=self.generator).tolist()
+                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=self._gen).tolist()
+            yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=self._gen).tolist()
         else:
-            yield from torch.randperm(n, generator=self.generator).tolist()
+            yield from torch.randperm(n, generator=self._gen).tolist()
+        self._gen = None
```
Comment (VitalyFedyunin) [inline torch/utils/data/sampler.py:128]: This will badly impact all currently running iterators, consider creating two individual iterators in test and yielding them in random order.

File: torch/utils/data/sampler.py [hunk 3]
```diff
@@ -112,15 +113,19 @@ def num_samples(self) -> int:
     def __iter__(self) -> Iterator[int]:
         n = len(self.data_source)
         if self.generator is None:
-            self.generator = torch.Generator()
-            self.generator.manual_seed(int(torch.empty((), dtype=torch.int64).random_().item()))
+            if self._gen is None:
```
Comment (VitalyFedyunin) [inline torch/utils/data/sampler.py:116]: Wait, can you please clarify what is going on here, it looks like you want individual instance of _gen for every iterator, but placing it into `self` not going to help.

**Review by VitalyFedyunin — COMMENTED** [review]
Not going to work as intended for two or more iterators created from same object.
```
 a = Sampler()
 i1 = iter(a)
 i2 = iter(a)
```

File: torch/utils/data/sampler.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/sampler.py:128]: Emmm, you are right. So, we have to create iter-local generator. But, then it goes back to the problem that we can not serialize the state of this iter-local generator anymore. Then, when we resume from snapshotting, we can not really fast forward the state of generator, have to iterate over it to the certain amount of iterations.

**Comment by VitalyFedyunin** [comment]
Return new generator from `__iter__` to keep states shared

```
class Sampler():
    def __init__(self):
        self.iterators = {}
        pass

    def __iter__(self):
        # Everything is by pointer here
        state_dict = {}
        generator = iter(self.generator(state_dict))
        self.iterators[generator] = state_dict
        return generator

    def generator(self, state_dict):
        state_dict['gen'] = len(self.iterators)
        for i in range(10):
            yield i

s = Sampler()
i = iter(s)
i2 = iter(s)

print(i)
print(i2)

for d in i:
    print(d)

print(s.iterators)
```

**Comment by ejguan** [comment]
> Return new generator from `__iter__` to keep states shared

Yeah, that's also something in my mind. Will update PR.

**Comment by ejguan** [comment]
Have to revert it back because `torch.Generator` is still not pickable for now. Need to fix it first.

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

**Comment by ejguan** [comment]
I split the original PR into two PRs.
This PR would revert Sampler back to previous state that the RNG is created lazily.

The another PR provide a new feature for potential users to serialize the iter-local generator.

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30451774).

**Review by NivekT — APPROVED** [review]
LGTM
